### PR TITLE
Fix performance of Foldable instance of streams for GHC-9

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Benchmark comparison for Prelude.Serial/o-1-space.foldable
**Without fix in GHC 9.0.1:**
minimum:                                                           OK
2.0 ms ± 2.8 ms, 9.9 MB allocated,  68 KB copied

maximum:                                                           OK (0.01s)
3.5 ms ± 2.9 ms,  13 MB allocated,  84 KB copied

**With oneShot fix in GHC 9.0.1::**
minimum:                                                           OK (0.01s)
 79 μs ± 206 μs,   0 B  allocated,   0 B  copied

maximum:                                                           OK (0.01s)
 99 μs ± 179 μs,   0 B  allocated,   0 B  copied

**Review by harendra-kumar — APPROVED** [review]
Please squash the commits.

**Comment by harendra-kumar** [comment]
@rnjtranjan need to add a CPP conditional for the 8.0.2 CI failure.

**Comment by rnjtranjan** [comment]
It's giving parse error like 
_src/Streamly/Internal/Data/Stream/Serial.hs:247:868: error:
    parse error on input ‘where’
    |
247 | FOLDABLE_INSTANCE(SerialT)_

when I put :

#if MIN_VERSION_base(4,15,0)                                                  \    
        where {f' x k = oneShot $ \z -> k $! f z x};                          \
#else                                                                                          \ 
        where {f' x k z = k $! f z x};                                                \
#endif                                                                                       \

I tried many diff combination with no success.

**Comment by harendra-kumar** [comment]
You can define `oneShot` in `Streamly.Internal.BaseCompat`. For older versions you can define it as `id`.

**Comment by rnjtranjan** [comment]
Fix #1087

File: src/Streamly/Internal/BaseCompat.hs [hunk 1]
```diff
@@ -55,3 +57,10 @@ unsafeWithForeignPtr = GHCForeignPtr.unsafeWithForeignPtr
 #else
 unsafeWithForeignPtr = withForeignPtr
 #endif
+
+oneShot :: (a -> b) -> a -> b
+#if MIN_VERSION_base(4,15,0)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/BaseCompat.hs:62]: Our build was failing only for 8.0, which means in other version onseShot is present. So we should use min version of base that was shipped with 8.2.

**Pull request merged**